drivers/sensor/mcp98xx/mcp98xx.c

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2019 Peter Bigot Consulting, LLC
+ * Copyright (c) 2016 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT microchip_mcp98xx
+
+#include <errno.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/init.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/logging/log.h>
+
+#include "mcp98xx.h"
+
+LOG_MODULE_REGISTER(MCP98XX, CONFIG_SENSOR_LOG_LEVEL);
+
+int mcp98xx_reg_read(const struct device *dev, uint8_t reg, uint16_t *val)
+{
+	const struct mcp98xx_config *cfg = dev->config;
+	int rc = i2c_write_read_dt(&cfg->i2c, &reg, sizeof(reg), val,
+				   sizeof(*val));
+
+	if (rc == 0) {
+		*val = sys_be16_to_cpu(*val);
+	}
+
+	return rc;
+}
+
+int mcp98xx_reg_write_16bit(const struct device *dev, uint8_t reg,
+			    uint16_t val)
+{
+	const struct mcp98xx_config *cfg = dev->config;
+
+	uint8_t buf[3];
+
+	buf[0] = reg;
+	sys_put_be16(val, &buf[1]);
+
+	return i2c_write_dt(&cfg->i2c, buf, sizeof(buf));
+}
+
+int mcp98xx_reg_write_8bit(const struct device *dev, uint8_t reg,
+			   uint8_t val)
+{
+	const struct mcp98xx_config *cfg = dev->config;
+	uint8_t buf[2] = {
+		reg,
+		val,
+	};
+
+	return i2c_write_dt(&cfg->i2c, buf, sizeof(buf));
+}
+
+bool mcp98xx_tconv_elapsed(uint8_t resulution, int64_t starttime) {
+	// Use typical values unless max is specified.
+	uint16_t tconv_time = 0;
+	switch(resulution) {
+		case 0b00: tconv_time = 30;  break; /// TYP
+		case 0b01: tconv_time = 120; break; /// MAX
+		case 0b10: tconv_time = 130; break; /// TYP
+		default:
+		case 0b11: tconv_time = 260; break; /// TYP
+	}
+	return k_uptime_delta(&starttime) > tconv_time;
+}
+
+static int mcp98xx_shutdown(const struct device *dev) {
+	uint16_t value = 0;
+	int result = mcp98xx_reg_read(dev, MCP98XX_REG_CONFIG, &value);
+	if (result == 0) {
+		value |= MCP98XX_REG_CONFIG_SHDN;
+		return mcp98xx_reg_write_16bit(dev, MCP98XX_REG_CONFIG, value );
+	}
+	return result;
+}
+
+static int mcp98xx_wakeup(const struct device *dev) {
+	uint16_t value = 0;
+	int result = mcp98xx_reg_read(dev, MCP98XX_REG_CONFIG, &value);
+	if (result == 0) {
+		value &= ~MCP98XX_REG_CONFIG_SHDN;
+		return mcp98xx_reg_write_16bit(dev, MCP98XX_REG_CONFIG, value );
+	}
+	return result;
+}
+
+static int mcp98xx_set_temperature_resolution(const struct device *dev,
+					      uint8_t resolution)
+{
+	int result;
+	uint16_t resolution_16bit = resolution;
+#if CONFIG_MCP98XX_CHIP_MCP9844
+	mcp98xx_shutdown(dev);
+#endif
+	result = mcp98xx_reg_write_16bit(dev, MCP98XX_REG_RESOLUTION, resolution_16bit);
```

Comment:
In the old MCP9808 code this was a call to `mcp9808_reg_write_8bit` and according to the datasheet this is indeed a 8 bit register in the MCP9808. Maybe make sure this is still set using the 8 bit function when the MCP9808 is selected.

Reply:
Solved, also no need to use `resolution_16bit` variable: it is copied on the stack